File: code/xproblem/CSharp/xproblem/visitorPattern/PrettyPrinter.cs
namespace visitorPattern
{
    public class PrettyPrinter : Visitor<string>
    {
        public string Visit(Const that) 
        { 
            return that.info.ToString(); 
        }
        public string Visit(Add that)
        {
            return that.left.Accept(this) + " + " + that.right.Accept(this); 
        }
    }
}

File: code/xproblem/Java/xproblem/visitorPattern/PrettyPrinter.java
package visitorPattern;

/**
 * Operation for pretty printing
 */
public class PrettyPrinter extends Visitor<String> {
	public String visit(Const x) {
			return Integer.toString(x.getInfo());
	}
	public String visit(Add x) {
			return x.getLeft().accept(this)
			     + " + "
			     + x.getRight().accept(this);
	}
	public String visit(Neg x) {
		return "- (" + x.getExpr().accept(this) + ")";
	}
}
